### DISCUSSION TITLE
Problem encountered IKVM 8.7.5 + H2 Database 2.2.220 + ojdbc8.jar + vs2019 + net472

### DISCUSSION THREAD

**Discussion by CoffeeKayGit** [ikvmnet/ikvm#488] [General]
Problem encountered IKVM 8.7.5 + H2 Database 2.2.220 + ojdbc8.jar + vs2019 + net472

I have created the h2.dll using IKVM 8.7.5 by merging the ojdbc8.jar

<ItemGroup>
	<IkvmReference Include="jar\ojdbc8.jar">
		<AssemblyName>o8</AssemblyName>
		<AssemblyVersion>1.0.0.8</AssemblyVersion>
	</IkvmReference>        
	<IkvmReference Include="jar\h2-2.2.220.jar">
		<AssemblyName>h2.2.2.220.8</AssemblyName>
		<AssemblyVersion>2.2.220.8</AssemblyVersion>
		<Compile>jar\ojdbc8.jar</Compile>
		<Reference>jar\ojdbc8.jar</Reference>
		<ClassLoader>ikvm.runtime.AppDomainAssemblyClassLoader</ClassLoader>
	</IkvmReference>
</ItemGroup>

using the generated dll file.

1. I have create a H2 database.
2. Connect to the database using connection string 
jdbc:h2:file:C:\WIP\ABC;USER=aaa;PASSWORD=aaa;ACCESS_MODE_DATA=r;IFEXISTS=TRUE;CASE_INSENSITIVE_IDENTIFIERS=TRUE
3. Create table DDD and insert some records then close the database.

there is no problem if connecting the database using readonly mode and select reocrd from table DDD

========================================================

Problem 1:
with read/write mode, the table DDD get deleted and never go back.
error: table DDD is not found.

=======================================================

Problem 2:
Cannot connect to another H2 database when creating linked table.

CREATE LOCAL  TEMPORARY  LINKED TABLE IF NOT EXISTS EEE (   
  'org.h2.Driver',   
  'jdbc:h2:file:C:/WIP/B1B199;USER=aaa;PASSWORD=aaaa;ACCESS_MODE_DATA=r;IFEXISTS=TRUE;CASE_INSENSITIVE_IDENTIFIERS=TRUE;DB_CLOSE_ON_EXIT=TRUE',
  'aaa',
  'aaaa',
  'DDD'
);

The error message is:

org.h2.jdbc.JdbcSQLNonTransientConnectionException: Database may be already in use: null. Possible solutions: close all other connection(s); use the 
server mode; SQL statement:
 CREATE LOCAL  TEMPORARY  LINKED TABLE IF NOT EXISTS EEE (   
'org.h2.Driver',   
'jdbc:h2:file:C:/WIP/B1B199;USER=aaa;PASSWORD=aaaa;ACCESS_MODE_DATA=r;IFEXISTS=TRUE;CASE_INSENSITIVE_IDENTIFIERS=TRUE;DB_CLOSE_ON_EXIT=TRUE',
'aaa',
'aaaa',
'DDD'  )
at org.h2.message.DbException.getJdbcSQLException(DbException.java:617)
at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
at org.h2.mvstore.db.MVTableEngine$Store.convertIllegalStateException(MVTableEngine.java:193)
at org.h2.mvstore.db.MVTableEngine$Store.open(MVTableEngine.java:173)
at org.h2.mvstore.db.MVTableEngine.init(MVTableEngine.java:96)
at org.h2.engine.Database.getPageStore(Database.java:2739)
at org.h2.engine.Database.open(Database.java:771)
at org.h2.engine.Database.openDatabase(Database.java:320)
at org.h2.engine.Database.<init>(Database.java:314)
at org.h2.engine.Engine.openSession(Engine.java:69)
at org.h2.engine.Engine.openSession(Engine.java:201)
at org.h2.engine.Engine.createSessionAndValidate(Engine.java:178)
at org.h2.engine.Engine.createSession(Engine.java:161)
at org.h2.engine.Engine.createSession(Engine.java:31)
at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:336)
at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:169)
at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:149)
at org.h2.Driver.connect(Driver.java:69)
at org.h2.util.JdbcUtils.getConnection(JdbcUtils.java:300)
at org.h2.util.JdbcUtils.getConnection(JdbcUtils.java:274)
at org.h2.table.TableLinkConnection.open(TableLinkConnection.java:89)
at org.h2.table.TableLinkConnection.open(TableLinkConnection.java:79)
at org.h2.engine.Database.getLinkConnection(Database.java:2708)
at org.h2.table.TableLink.connect(TableLink.java:96)
at org.h2.table.TableLink.<init>(TableLink.java:79)
at org.h2.schema.Schema.createTableLink(Schema.java:745)
at org.h2.command.ddl.CreateLinkedTable.update(CreateLinkedTable.java:77)
at org.h2.command.CommandContainer.update(CommandContainer.java:133)
at org.h2.command.Command.executeUpdate(Command.java:267)
at org.h2.jdbc.JdbcStatement.executeUpdateInternal(JdbcStatement.java:169)
at org.h2.jdbc.JdbcStatement.executeUpdate(JdbcStatement.java:126)

=============================================================

Problem 3:
Cannot connect to Oracle19c when creating linked table.

CREATE LOCAL TEMPORARY LINKED TABLE IF NOT EXISTS TABLEB (
'',  
'jdbc:oracle:thin:@(DESCRIPTION = (ADDRESS_LIST = (ADDRESS = (PROTOCOL = 
TCP)(HOST = 1.2.3.4)(PORT = 1630)) (ADDRESS = (PROTOCOL = TCP)(HOST = 
4.3.2.1)(PORT = 1531))) (SOURCE_ROUTE = yes) (CONNECT_DATA = (SERVICE_NAME 
= O19CDB)))',  
'userid',  
'password',  
'schema.TABLEA'  
) READONLY; 

The error message is:

General error: java.lang.IndexOutOfBoundsException (Error Type: org.h2.jdbc.JdbcSQLNonTransientException)
at org.h2.util.JdbcUtils.getconnection(String str1, String str2, String str3, String str4, NetworkConnectionInfo nci, Boolean b)
at org.h2.util.JdbcUtils.getconnection(String str1, String str2, String str3, String str4)
at org.h2.table.TableLinkConnection.open()

==========================================

I can do all the 3 steps using SQuirreL SQL.

**Comment by wasabii** [comment]
Well, without more information, it's a little hard to conclude this is an IKVM issue, and not just an issue with using the particular library.

**Comment by zznty** [comment]
ive also observed this issue with h2, it simply doesnt save anything to disk, while the same code being absolutely fine on jvm.

**Comment by CoffeeKayGit** [comment]
thanks. 

my project requires:
1. create a embedded h2 database
2. connect to it using read write mode
3. create linked table to oracle, grab some tables into the h2 database
4. on production, connect to H2_DB1 using read only mode, create linked table to another H2_DB2, do select query.

I can do this with IKVM 7.2.4630.5 + H2 1.4.199.jar + ojdbc6.jar + vs2019 + net472, but now need to upgrade.

**Comment by CoffeeKayGit** [comment]
I will provide detailed steps to repeat the error.

I have created 2 projects.
project 1 is the steps to generate dll

Step 1. Generate h2.dll
1. create .net sdk project ikvm.test.jar2dll, change to net472
2. add lib ikvm 8.7.5
3. download ojdbc.jar and h2-2.2.220.jar from official site
4. add IkvmReference
5. build.
6. The output file h2.dll can be generated.
[01_TestGenDLL.zip](https://github.com/ikvmnet/ikvm/files/14291155/01_TestGenDLL.zip)

=============================================

project 2 is the steps to repeat error

1. create a c# winform project with net472
2. copy and ref h2.dll
3. add lib package ikvm 8.7.5, but failed
    Unable to find a version of 'Mono.Posix' that is compatible with 'IKVM 8.7.5 constraint: Mono.Posix (>= 7.1.0-final.1.21458.1)'
4. so I use command to install Mono.Posix in Package Manager Console
    NuGet\Install-Package Mono.Posix -Version 7.1.0-final.1.21458.1
5. install ikvm 8.7.5 again, succeed.
6. copy folder ikvm and runtimes from previous project to this project
7. open the project
8. click the buttons step by step.
[02_RepeatError.zip](https://github.com/ikvmnet/ikvm/files/14291163/02_RepeatError.zip)

**Reply by wasabii** [reply]
So, quick observation here: IKVM (since this project took it over, so 2+ years now) dropped support for old-style MSBuild projects. We require new-style SDK projects. All of the packages and build scripts are designed with that expectation. Which could relate to why things aren't going right for you. For instance, the project you sent: none of the required IKVM libraries are present in the output. No native ikvm.dll file. No ikvm directory with the JVM native libraries (nio.dll, net.dll, etc). All of these are part of the NuGet packages, which won't restore correctly because you are on a old-style project file.

I'm also super confused why you are using a separate project to generate a DLL only to copy it into the final project. Just add IkvmReference to the project you're using it from.

https://github.com/ikvmnet/ikvm?tab=readme-ov-file#support

**Reply by CoffeeKayGit** [reply]
thanks for the reply.

for project 2
I have installed ikvm 8.7.5 too and
in step 6, copied ikvm folder and runtimes folder from project 1 into project 2 bin folder.
that's why I included a text file for listing all files inside these 2 folders in the zip file of project 1.
(these 2 folders are too larger for uploading, I have deleted them before uploading.)

it may be the root cause as you said.
I can do another test by merging everything into a single .net sdk project.

**Comment by CoffeeKayGit** [comment]
More things to say:
1. I can do all steps without error using SQuirreL SQL
2. h2-2.2.220.jar is chosen instead of latest version. It seems JDK8u92-b34 does not support latest H2
3. In past, I chose to merge ojdbc6.jar into the built h2.1.4.199.dll, so that H2 will use ojdbc properly.
4. the possible root cause may be...
a. need to do anything with java class_path or java environment variable?
b. need to do anything with java class loader?
c. the dll is build as x64, but the project run as x32, any problem with this?
d. set application base path?
e. for error "database is locked", some said h2 embedded mode support only 1 connection, need to make sure there is only 1 jvm creating connection to the database, but how to do so?
f. any mistake when generating the dll?
g. when h2 try to create a temp file in %AppData%\Local\Temp\, the temp file is locked. (doing ORDER BY or CREATE LINKED TABLE)
h. need to install package ikvm.image.jre.runtime?
